src/winconpty/winconpty.cpp

```diff
@@ -525,13 +538,14 @@ extern "C" HRESULT WINAPI ConptyShowHidePseudoConsole(_In_ HPCON hPC, bool show)
 // - Used to support GH#2988
 extern "C" HRESULT WINAPI ConptyReparentPseudoConsole(_In_ HPCON hPC, HWND newParent)
 {
-    const PseudoConsole* const pPty = (PseudoConsole*)hPC;
-    auto hr = pPty == nullptr ? E_INVALIDARG : S_OK;
-    if (SUCCEEDED(hr))
-    {
-        hr = _ReparentPseudoConsole(pPty, newParent);
-    }
-    return hr;
+    return _ReparentPseudoConsole((PseudoConsole*)hPC, newParent);
+}
+
+// - Sends a message to the pseudoconsole informing it that it should
+//   exit as soon as all console clients have disconnected.
+extern "C" HRESULT WINAPI ConptySetPseudoConsoleAutoExit(_In_ HPCON hPC)
+{
+    return _SetPseudoConsoleAutoExit((PseudoConsole*)hPC);

```

Comment:
#4564 makes a good point about:
> On POSIX systems the master end will receive an EOF when all processes that use the slave have terminated or closed the connection.

...should we maybe make auto-exit the default? It might however make a session reconnect feature a bit more difficult in the future (since you might not want OpenConsole to exit in that case).

BTW this is a good point to talk about why I made this new API the way it is: An alternative approach would've been to allow a user to specify an event handle, which OpenConsole could signal whenever all clients have disconnected. A ConPTY user could then choose to call `ClosePseudoConsole`. But I'm very confident in saying that this is probably not what most people want from ConPTY. Instead the above POSIX behavior seems way simpler to use: Just read from your pipes until the pipes are EOF. Boom, done. In fact this is even sufficient for WT, since we don't have a session reconnect feature yet either.

Reply:
> It might however make a session reconnect feature a bit more difficult in the future (since you might not want OpenConsole to exit in that case).

I'm just entirely instantiating a new ConptyConnection entirely, including a new openconsole, when the user requests a connection restart. So that I'm not too worried about.

Reply:
POSIX PTYs are kernel objects that are neither running nor exited. By default, if you are holding one end of a PTY, it stays open. That's the opposite of your assertion!

Reply:
I can't find this behavior documented in any manual page ala `openpty`, `pty(7)`, `ptmx`, etc.

Reply:
While I couldn't reproduce this with `openpty` either, I found that `forkpty` does indeed work exactly like they described it: Once the last PTY client has exited, `read()` will continue returning -1 and `errno` will be set to `EIO`. I suspect this might have to do with it using `login_tty`...